federation/pkg/kubefed/join.go

```diff
@@ -0,0 +1,319 @@
+	join_example = templates.Examples(`
+		# Join a cluster to a federation by specifying the
+		# cluster context name and the context name of the
+		# federation control plane's host cluster.
+		kubectl join foo --host=bar`)
+)
+
+// JoinFederationConfig provides a filesystem based kubeconfig (via
+// `PathOptions()`) and a mechanism to talk to the federation host
+// cluster.
+type JoinFederationConfig interface {
+	// PathOptions provides filesystem based kubeconfig access.
+	PathOptions() *clientcmd.PathOptions
+	// HostFactory provides a mechanism to communicate with the
+	// cluster where federation control plane is hosted.
+	HostFactory(host, kubeconfigPath string) cmdutil.Factory
+}
+
+// joinFederationConfig implements JoinFederationConfig interface.
+type joinFederationConfig struct {
+	pathOptions *clientcmd.PathOptions
+}
+
+func NewJoinFederationConfig(pathOptions *clientcmd.PathOptions) JoinFederationConfig {
+	return &joinFederationConfig{
+		pathOptions: pathOptions,
+	}
+}
+
+func (r *joinFederationConfig) PathOptions() *clientcmd.PathOptions {
+	return r.pathOptions
+}
+
+func (r *joinFederationConfig) HostFactory(host, kubeconfigPath string) cmdutil.Factory {
+	loadingRules := *r.pathOptions.LoadingRules
+	loadingRules.Precedence = r.pathOptions.GetLoadingPrecedence()
+	loadingRules.ExplicitPath = kubeconfigPath
+	overrides := &clientcmd.ConfigOverrides{
+		CurrentContext: host,
+	}
+
+	hostClientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(&loadingRules, overrides)
+
+	return cmdutil.NewFactory(hostClientConfig)
+}
+
+// NewCmdJoin defines the `join` command that joins a cluster to a
+// federation.
+func NewCmdJoin(f cmdutil.Factory, cmdOut io.Writer, config JoinFederationConfig) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "join CLUSTER_CONTEXT --host=HOST_CONTEXT",
+		Short:   "Join a cluster to a federation",
+		Long:    join_long,
+		Example: join_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := joinFederation(f, cmdOut, config, cmd, args)
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	cmdutil.AddApplyAnnotationFlags(cmd)
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	cmdutil.AddGeneratorFlags(cmd, cmdutil.ClusterV1Beta1GeneratorName)
+	addJoinFlags(cmd)
+	return cmd
+}
+
+// joinFederation is the implementation of the `join federation` command.
+func joinFederation(f cmdutil.Factory, cmdOut io.Writer, config JoinFederationConfig, cmd *cobra.Command, args []string) error {
+	name, err := kubectlcmd.NameFromCommandArgs(cmd, args)
+	if err != nil {
+		return err
+	}
+	host := cmdutil.GetFlagString(cmd, "host")
+	hostSystemNamespace := cmdutil.GetFlagString(cmd, "host-system-namespace")
+	kubeconfig := cmdutil.GetFlagString(cmd, "kubeconfig")
+	dryRun := cmdutil.GetDryRunFlag(cmd)
+
+	po := config.PathOptions()
+	po.LoadingRules.ExplicitPath = kubeconfig
+	clientConfig, err := po.GetStartingConfig()
+	if err != nil {
+		return err
```

Comment:
Its generally recommended to add some context while returning error message. Like
`return fmt.Errorf("error %s in getting starting config for path options: %v", po, err)`